app/src/main/java/me/hufman/androidautoidrive/carapp/music/views/EnqueuedView.kt

```diff
@@ -1,77 +1,219 @@
-		state.getTextModel()?.asRaDataModel()?.value = L.MUSIC_QUEUE_TITLE
+		// this is required for pagination system
+		listComponent.setProperty(RHMIProperty.PropertyId.VALID, false)
+
 		listComponent.setVisible(true)
-		listComponent.setProperty(RHMIProperty.PropertyId.LIST_COLUMNWIDTH, "57,50,*")
+		listComponent.setProperty(RHMIProperty.PropertyId.LIST_COLUMNWIDTH, "57,90,10,*")
 		listComponent.getAction()?.asHMIAction()?.getTargetModel()?.asRaIntModel()?.value = playbackView.state.id
 		listComponent.getAction()?.asRAAction()?.rhmiActionCallback = RHMIActionListCallback { onClick(it) }
+		listComponent.getSelectAction()?.asRAAction()?.rhmiActionCallback = RHMIActionListCallback { onSelectAction(it) }
 	}
 
 	fun show() {
 		currentSong = musicController.getMetadata()
+		val newQueueMetadata = musicController.getQueue()
+
+		// same queue as before, just select currently playing song
+		if (queueMetadata != null && queueMetadata == newQueueMetadata) {
+			showCurrentlyPlayingSong()
+			return
+		}
+
+		queueMetadata = newQueueMetadata
 		songsList.clear()
-		val songs = musicController.getQueue()
+		val songs = queueMetadata?.songs
 		if (songs?.isNotEmpty() == true) {
 			listComponent.setEnabled(true)
 			listComponent.setSelectable(true)
 			songsList.addAll(songs)
-			listComponent.getModel()?.setValue(songsListAdapter, 0, songsListAdapter.height, songsListAdapter.height)
-
-			// set the selection to the current song
-			val index = songs.indexOfFirst { it.queueId == currentSong?.queueId }
-			if (index >= 0) {
-				state.app.events.values.firstOrNull { it is RHMIEvent.FocusEvent }?.triggerEvent(
-						mapOf(0.toByte() to listComponent.id, 41.toByte() to index)
-				)
+
+			listComponent.requestDataCallback = RequestDataCallback { startIndex, numRows ->
+				showList(startIndex, numRows)
+
+				val endIndex = if (startIndex + numRows >= songsList.size) songsList.size - 1 else startIndex + numRows
+				visibleRows = songsListAdapter.realData.subList(startIndex, endIndex + 1)
+				visibleRowsOriginalMusicMetadata.clear()
+				visibleRows.forEach { musicMetadata ->
+					visibleRowsOriginalMusicMetadata.add(MusicMetadata.copy(musicMetadata))
+				}
 			}
+
+			showCurrentlyPlayingSong()
 		} else {
 			listComponent.setEnabled(false)
 			listComponent.setSelectable(false)
 			listComponent.getModel()?.setValue(songsEmptyList, 0, songsEmptyList.height, songsEmptyList.height)
 		}
+
+		val queueTitle = UnicodeCleaner.clean(queueMetadata?.title ?: "")
+		val queueSubtitle = UnicodeCleaner.clean(queueMetadata?.subtitle ?: "")
+		if (queueTitle.isBlank() && queueSubtitle.isBlank()) {
+			state.getTextModel()?.asRaDataModel()?.value = "Now Playing"
+		} else {
+			state.getTextModel()?.asRaDataModel()?.value = "$queueTitle - $queueSubtitle"
+		}
+
+		if (queueMetadata?.coverArt != null) {
+			queueImageComponent.getModel()?.asRaImageModel()?.value = graphicsHelpers.compress(musicController.getQueue()?.coverArt!!, 180, 180, quality = 60)
```

Comment:
The `musicController.getQueue()` could theoretically change after the null check. I'd recommend saving `queueMetadata?.coverArt` to a local variable and doing the null check on that local variable, then reusing it for this compress call.

Reply:
Good point, changed this to be stored in a local variable